src/System.Private.Xml/src/System/Xml/XmlDownloadManager.cs

```diff
@@ -36,22 +36,242 @@ internal partial class XmlDownloadManager
         private Stream GetNonFileStream(Uri uri, ICredentials credentials, IWebProxy proxy,
             RequestCachePolicy cachePolicy)
         {
-            HttpClient client = new HttpClient();
-            HttpRequestMessage req = new HttpRequestMessage(HttpMethod.Get, uri);
+            WebRequest req = WebRequest.Create(uri);
+            if (credentials != null)
+            {
+                req.Credentials = credentials;
+            }
+            if (proxy != null)
+            {
+                req.Proxy = proxy;
+            }
+            if (cachePolicy != null)
+            {
+                req.CachePolicy = cachePolicy;
+            }
+            WebResponse resp = req.GetResponse();
+            HttpWebRequest webReq = req as HttpWebRequest;
+            if (webReq != null)
+            {
+                lock (this)
+                {
+                    if (_connections == null)
+                    {
+                        _connections = new Hashtable();
+                    }
+                    OpenedHost openedHost = (OpenedHost)_connections[webReq.Address.Host];
+                    if (openedHost == null)
+                    {
+                        openedHost = new OpenedHost();
+                    }
 
+                    if (openedHost.nonCachedConnectionsCount < webReq.ServicePoint.ConnectionLimit - 1)
+                    {
+                        // we are not close to connection limit -> don't cache the stream
+                        if (openedHost.nonCachedConnectionsCount == 0)
+                        {
+                            _connections.Add(webReq.Address.Host, openedHost);
+                        }
+                        openedHost.nonCachedConnectionsCount++;
+                        return new XmlRegisteredNonCachedStream(resp.GetResponseStream(), this, webReq.Address.Host);
+                    }
+                    else
+                    {
+                        // cache the stream and save the connection for the next request
+                        return new XmlCachedStream(resp.ResponseUri, resp.GetResponseStream());
+                    }
```

Comment:
Are there tests that exercise this caching?  I would not expect this to behave correctly given the HttpWebRequest implementation on top of HttpClient.